### PR TITLE
go/registry: Optimize check for duplicate keys

### LINKED ISSUE #2454: go/registry: Optimize check for duplicate keys on node registration
On new node registration, we currently go through all registered nodes to check that the new node's consensus and P2P keys are unique, which is not very optimal.
It would be better to have a map of keys to node IDs, so that we can check if a key already exists with a simple lookup.

### PR DESCRIPTION
**Pull request by abukosek**
Closes #2454.

### CODE REVIEW

File: go/consensus/tendermint/apps/registry/state/state.go [hunk 1]
```diff
@@ -50,6 +51,16 @@ var (
...
+	// keyMapKeyFmt is the key format used for key-to-node-id map.
+	// This stores the consensus and P2P to Node ID mappings.
+	//
+	// Value is binary signature.PublicKey (node ID).
+	keyMapKeyFmt = keyformat.New(0x17, &signature.PublicKey{})
+	// certificateMapKeyFmt is the key format used for certificate-to-node-id map.
+	// This stores the hash-of-certificate to Node ID mappings.
+	//
+	// Value is binary signature.PublicKey (node ID).
+	certificateMapKeyFmt = keyformat.New(0x18, []byte{})
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/state/state.go]: Why `[]byte{}` instead of `&hash.Hash{}`?

Comment (abukosek) [reply go/consensus/tendermint/apps/registry/state/state.go]: Oops, fixed! :)

File: go/consensus/tendermint/apps/registry/state/state.go [hunk 2]
```diff
@@ -390,6 +401,40 @@ func (s *ImmutableState) ConsensusParameters() (*registry.ConsensusParameters, e
 	return &params, err
 }
 
+func (s *ImmutableState) NodeByConsensusOrP2PKey(key signature.PublicKey) (*node.Node, error) {
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/state/state.go]: Add type assertion at the top that `ImmutableState` implements `NodeLookup`.

**Comment by abukosek** [comment]
Thanks!

**Pull request merged**